cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterCreator.java

```diff
@@ -0,0 +1,267 @@
+    private final boolean useKRaftFGEnabled;
+    private final KafkaVersion.Lookup versions;
+
+    // Operators and other tools
+    private final Vertx vertx;
+    private final AdminClientProvider adminClientProvider;
+    private final SecretOperator secretOperator;
+    private final SharedEnvironmentProvider sharedEnvironmentProvider;
+    private final PreventBrokerScaleDownCheck brokerScaleDownOperations;
+
+    // State
+    private boolean scaleDownCheckFailed = false;
+
+    /**
+     * Constructor
+     *
+     * @param vertx             Vert.x instance
+     * @param reconciliation    Reconciliation marker
+     * @param config            Cluster Operator configuration
+     * @param supplier          Resource Operators supplier
+     */
+    public KafkaClusterCreator(
+            Vertx vertx,
+            Reconciliation reconciliation,
+            ClusterOperatorConfig config,
+            ResourceOperatorSupplier supplier
+    ) {
+        this.reconciliation = reconciliation;
+        this.versions = config.versions();
+        this.useKRaftFGEnabled = config.featureGates().useKRaftEnabled();
+
+        this.vertx = vertx;
+        this.adminClientProvider = supplier.adminClientProvider;
+        this.secretOperator = supplier.secretOperations;
+        this.sharedEnvironmentProvider = supplier.sharedEnvironmentProvider;
+        this.brokerScaleDownOperations = supplier.brokerScaleDownOperations;
+    }
+
+    /**
+     * Prepares the Kafka Cluster model instance. It checks if any scale-down is happening and whether such scale-down
+     * can be done. If it discovers any problems, it will try to fix them and create a fixed Kafka Cluster model
+     * instance if the {@code tryToFixProblems} fag is set to true. It throws an exception otherwise.
+     *
+     * @param kafkaCr           Kafka custom resource
+     * @param nodePools         List with Kafka Node Pool resources
+     * @param oldStorage        Old storage configuration
+     * @param currentPods       Existing Kafka pods
+     * @param versionChange     Version Change object describing any possible upgrades / downgrades
+     * @param tryToFixProblems  Flag indicating whether recoverable configuration issues should be fixed or not
+     *
+     * @return  New Kafka Cluster instance
+     */
+    public Future<KafkaCluster> prepareKafkaCluster(
+            Kafka kafkaCr,
+            List<KafkaNodePool> nodePools,
+            Map<String, Storage> oldStorage,
+            Map<String, List<String>> currentPods,
+            KafkaVersionChange versionChange,
+            boolean tryToFixProblems
+    )   {
+        return createKafkaCluster(kafkaCr, nodePools, oldStorage, currentPods, versionChange)
+                .compose(kafka -> brokerScaleDownCheck(kafkaCr, kafka))
+                .compose(kafka -> {
+                    if (scaleDownCheckFailed && tryToFixProblems)   {
+                        // We have a failure, and should try to fix issues
```

Comment:
isn't this comment inconsistent with `false` you are passing later on to the prepare? Or am I missing the meaning of it?

Reply:
It is first called with `tryToFixProblems` set to `true` when called from `KafkaAssemblyOperator`. If we run into any issues in the first run, we try to fix them and call it again, but this time with `tryToFixProblems` set to `false` so that we recreate the `KafkaCluster` object. But if it is still failing, then we just throw up the exception -> so that if there is some issue we cannot fix we don't loop forever.

Reply:
I think I misleading the comment. It was referring to the "if" branch we are in (so as you said try fix problems is true) while I was looking at how prepare is called later passing false (so you are not trying to fix problems).
Thanks.

Reply:
I will add a comment to make it more clear.